pkg/balancer/least_connections.go

```diff
@@ -0,0 +1,129 @@
+package balancer
+
+import (
+	"math"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/square/p2/pkg/util"
+)
+
+type lcBackend struct {
+	conns int
+
+	// disabling allows us to gracefully release connections without assigning
+	// new connections to the backend.
+	disabled bool
+}
+
+type LeastConnections struct {
+	backends map[string]*lcBackend
+	mapMux   sync.Mutex
+}
+
+func NewLeastConnectionsStrategy(initialAddresses []string) (*LeastConnections, error) {
+	l := &LeastConnections{backends: make(map[string]*lcBackend)}
+	for _, addr := range initialAddresses {
+		err := l.AddAddress(addr)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return l, nil
+}
+
+func (l *LeastConnections) Route(fn Handler) error {
+	address, err := l.acquireAddress()
+	if err != nil {
+		return util.Errorf("Could not acquire %s: %s", address, err)
+	}
+	defer l.releaseAddress(address)
+	outbound, err := net.Dial("tcp", address)
+	if err != nil {
+		return util.Errorf("Could not dial %s: %s", address, err)
+	}
+	defer outbound.Close()
+	err = fn(outbound.(*net.TCPConn))
+	if err != nil {
+		return util.Errorf("Could not handle connection to %s: %s", address, err)
+	}
+	return nil
+}
+
+func (l *LeastConnections) RemoveAddress(address string) error {
+	l.mapMux.Lock()
+	defer l.mapMux.Unlock()
+	backend, ok := l.backends[address]
+	if !ok {
+		return nil
+	}
+	backend.disabled = true
```

Comment:
Here and in `releaseAddress()` below, once the record becomes disabled with no connections, it should eventually be removed from `backends` to prevent a memory leak.